listings/applications/erc721/src/tests.cairo

```diff
@@ -0,0 +1,293 @@
+    assert(uri == '', 'wrong uri');
+
+    contract.set_token_uri(nft_id, 'https://example.com/1');
+    let uri = contract.get_token_uri(nft_id);
+    assert(uri == 'https://example.com/1', 'wrong uri');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_transfer_from() {
+    let (project_address, _) = deploy_project('ApeMonkey', 'APE');
+    let contract = IERC721Dispatcher{ contract_address: project_address };
+
+    let OWNER: ContractAddress = contract_address_const::<'OWNER'>();
+    start_prank(CheatTarget::One(project_address), OWNER);
+
+    let user = starknet::contract_address_const::<'USER'>();
+    let nft_id: u256 = 1.into();
+    contract.mint(OWNER, nft_id);
+
+    contract.transfer_from(OWNER, user, nft_id);
+
+    let new_owner = contract.owner_of(nft_id);
+    assert(new_owner == user, 'wrong new OWNER');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_transfer_from_approved() {
+    let (project_address, _) = deploy_project('ApeMonkey', 'APE');
+    let contract = IERC721Dispatcher{ contract_address: project_address };
+    let OWNER: ContractAddress = contract_address_const::<'OWNER'>();
+    start_prank(CheatTarget::One(project_address), OWNER);
+
+    let user = starknet::contract_address_const::<'USER'>();
+    let nft_id: u256 = 1.into();
+
+    contract.mint(OWNER, nft_id);
+    contract.approve(user, nft_id);
+
+    contract.transfer_from(OWNER, user, nft_id);
+
+    let new_owner = contract.owner_of(nft_id);
+    assert(new_owner == user, 'wrong new OWNER');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_transfer_from_approved_for_all() {
+    let (project_address, _) = deploy_project('ApeMonkey', 'APE');
+    let contract = IERC721Dispatcher{ contract_address: project_address };
+
+    let OWNER: ContractAddress = contract_address_const::<'OWNER'>();
+    start_prank(CheatTarget::One(project_address), OWNER);
+    
+    let user = starknet::contract_address_const::<'USER'>();
+
+    let nft_id: u256 = 1.into();
+    contract.mint(OWNER, nft_id);
+    contract.set_approval_for_all(user, true);
+
+    contract.transfer_from(OWNER, user, nft_id);
+
+    let new_owner = contract.owner_of(nft_id);
+    assert(new_owner == user, 'wrong new OWNER');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic]
+fn test_transfer_from_not_approved() {
+
+    let (project_address, _) = deploy_project('ApeMonkey', 'APE');
+    let contract = IERC721Dispatcher{ contract_address: project_address };
+
+    let OWNER: ContractAddress = contract_address_const::<'OWNER'>();
+    start_prank(CheatTarget::One(project_address), OWNER);
+
+    let user = starknet::contract_address_const::<'USER'>();
+    let nft_id: u256 = 1.into();
+
+    contract.mint(OWNER, nft_id);
+    contract.approve(user, nft_id);
+
+    let random_user = starknet::contract_address_const::<789>();
+    contract.transfer_from(random_user, user, nft_id);
+
+    let new_owner = contract.owner_of(nft_id);
+    assert(new_owner == user, 'wrong new OWNER');
+}
```

Comment:
Still missing the newline character, please resolve the conversation only once the request is addressed.